src/System.Private.CoreLib/shared/System/Number.NumberToDouble.cs

```diff
@@ -0,0 +1,450 @@
+
+        private static void CheckTables()
+        {
+            ulong val;
+            int exp;
+
+            val = 0xa000000000000000;
+            exp = 4; // 10
+
+            CheckPow10MantissaTable(val, exp, new Span<ulong>(rgval64Power10, 0, 15), "rgval64Power10");
+            CheckPow10ExponentTable(val, exp, new Span<short>(rgexp64Power10, 0, 15), "rgexp64Power10");
+
+            val = 0x8E1BC9BF04000000;
+            exp = 54; //10^16
+
+            CheckPow10MantissaTable(val, exp, new Span<ulong>(rgval64Power10By16, 0, 21), "rgval64Power10By16");
+            CheckPow10ExponentTable(val, exp, new Span<short>(rgexp64Power10By16, 0, 21), "rgexp64Power10By16");
+
+            val = 0xCCCCCCCCCCCCCCCD;
+            exp = -3; // 0.1
+
+            CheckPow10MantissaTable(val, exp, new Span<ulong>(rgval64Power10, 15, 15), "rgval64Power10 - inv");
+
+            val = 0xE69594BEC44DE160;
+            exp = -53; // 0.1^16
+
+            CheckPow10MantissaTable(val, exp, new Span<ulong>(rgval64Power10By16, 21, 21), "rgval64Power10By16 - inv");
+        }
+
+        // debug-only verification of the precomputed tables
+        private static void CheckPow10MantissaTable(ulong val, int exp, Span<ulong> table, string name)
+        {
+            ulong multval = val;
+            int mulexp = exp;
+            bool isBad = false;
+
+            for (int i = 0; i < table.Length; i++)
+            {
+                if (table[i] != val)
+                {
+                    if (!isBad)
+                    {
+                        Debug.WriteLine(name);
+                        isBad = true;
+                    }
+                    Debug.WriteLine($"0x{val:X16},     // {i + 1}");
+                }
+
+                exp += mulexp;
+                val = Mul64Precise(val, multval, ref exp);
+            }
+
+            Debug.Assert(!isBad, "NumberToDouble table not correct. Correct version dumped to debug output.");
+        }
+
+        // debug-only verification of the precomputed tables
+        private static void CheckPow10ExponentTable(ulong val, int exp, Span<short> table, string name)
+        {
+            ulong multval = val;
+            int mulexp = exp;
+            bool isBad = false;
+
+            for (int i = 0; i < table.Length; i++)
+            {
+                if (table[i] != exp)
+                {
+                    if (!isBad)
+                    {
+                        Debug.WriteLine(name);
+                        isBad = true;
+                    }
+                    Debug.WriteLine($"{val}, // {i + 1}");
+                }
+
+                exp += mulexp;
+                val = Mul64Precise(val, multval, ref exp);
+            }
+
+            Debug.Assert(!isBad, "NumberToDouble table not correct. Correct version dumped to debug output.");
+        }
+#endif
+
+        // get 32-bit integer from at most 9 digits
+        private static uint DigitsToInt(char* p, int count)
```

Comment:
why not 10? (2147483647 is 10 digits)

Reply:
because the 10th digit can only represent 0-4, it cannot represent 5-9